### ISSUE TITLE
Track incremental perf improvements / refactorings in `RegularGridInterpolator(..., method="linear")`

### ISSUE THREAD

**Issue by ev-br** [scipy/scipy#17356]
https://github.com/scipy/scipy/pull/17291 implemented a fast path for the 2D linear mode of the RegularGridInterpolator. That is an initial implementation, and this issue is to keep track of possible incremental improvements.

First, several possible experiments to see if the current implementation (see `interpolate/_rgi.py` and `interpolate/_rgi_cython.pyx`) can be further sped up or streamlined:

- [ ] experiment with the `tuple grid` argument in cython: try to either pad the individual elements (so that the grid becomes an N-D array) and pass around the real sizes, or bite the bullet and switch over to `double**` (maybe not this though). This would allow to remove last bits of python interaction from cython code, the latter would then be possible to be declare `nogil`. Will need to see if this helps with perf or not.
- [ ]  see if we can require memoryviews to be C contiguous. This would require I think to loop over the trailing dimensions of values explicitly (not a big deal if somewhat messy) and maybe get the transposes right to make sure that cython loops respect memory locality. Also this would require some np.ascontiguousarrays, which might involve copying, so there are memory considerations. Again, the impact will need to be measured.
- [ ] see if we can simplify the cython logic: ATM both size-1 axes and nan inputs make find_indices return -1. Maybe we could simplify that. One complication is that the current fast path cannot fully replace python-level evaluate_linear because of "duck-typed arrays" (looking at the original PR from 2014 which introduced RGI, these are meant to take care of out-of-memory arrays like NetCDF files etc).

Additional points brought up in gh-17291 review:

- [ ] Consider adding a 3D fast path.
- [x] Consider adding a generic N-D cythonized evaluation. This will hopefully obviate the need for fast paths, but so far it seems to be difficult to loop over runtime-specified number of dimensions in Cython. In a nutshell, this needs a performant no-python  `itertools.product` replacement.

An alternative if quite bigger project would be to implement an `NdBSpline` class to mirror `NdPPoly`. 
This has a potential to significantly improve both linear and spline RGI modes: the former could then be replaced by an "slinear" mode , and the latter would no longer need recursive `make_interp_spline` calls. 

Update 02/2024: https://github.com/scipy/scipy/pull/19633 has been merged in the 1.13 release cycle, so two last items (an N-D evaluation and NdBSplin) are done, and are available as `mode="slinear"`

**Comment by notstarboard** [comment]
> Thanks for sharing your perspective. Much appreciated!
> 
> Re: porting guide: I get the feeling that you haven't seen https://gist.github.com/ev-br/8544371b40f414b7eaf3fe6217209bff which lists precise fitpack equivalents (FITPACK is the Fortran library from the 80s which interp2d delegates the work to). So the questions for the porting guide are
> 
>     * is this gist clear? If not, how to improve it? Ideally, send a pull request; if that's hard, can put the desired changes here.
> 
>     * how this gist could be made more discoverable? It currently is in the deprecation message, the link is straight from the red box in https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp2d.html
>       Suggestions to where to put the link so that users see it are greatly appreciated
> 
> 
> Also note that for the time being you can just silence the DeprecationWarning. Given that interp2d is indeed quite widely used, the plan is to have an extended deprecation period. But silencing the warning I personally would not recommend, better follow the equivalence pointed out in that gist above.

Referring to this post from issue #18010, there are a few things that should be improved. First, the red box in the scipy docs suggests to use the RegularGridInterpolator, but the performance vs interp2d makes it actually unusable for anything substantial. Second, the link to the gist is not clickable in the box (none of the links are), and the URL contains a hodgepodge of random letters and numbers, so nobody is going to navigate to that unless they're very confused. I didn't know what a gist even was until reading through issue #18010, and I'm sure I'm not the only ignorant one out there.

At any rate, when I see a recommendation to use option A for legacy code and option B for new code, what that implies to me is: "option B is actually the recommended option going forward, and you will likely have to switch over to it eventually, but option A will be easier to implement as a direct replacement". So, I initially changed my code to use RegularGridInterpolator, but that was clearly a mistake given its performance issues.

I understand there are concerns with maintaining interp2d and I don't have any problem with prioritizing making code easy to maintain. I'm sure you have compelling reasons to deprecate it. But until RegularGridInterpolator is at least comparable to interp2d in terms of performance, the following should be done:

1) The RegularGridInterpolator documentation should be given a big red warning at the top to explain that performance is substantially worse than interp2d, and that users should consider the options outlined on the gist page in the likely event that performance is a concern.
2) RegularGridInterpolator should not be recommended for new code at the top of interp2d unless there is a clear caveat there about the poor performance. Knowing now how bad the performance is, I can't even consider using RegularGridInterpolation for any of my code, new or legacy. It's unusuable for any heavy lifting in its current state, even with linear interpolation, and users should not have to waste the time necessary to figure that out the hard way. 

Apologies if this is the wrong place for this feedback; I didn't want to drop this on #18010 since that has been closed. Thanks!

**Comment by j-bowhay** [comment]
Apologies about the link not being clickable, it is an upstream issue with the theme that is causing us a whole world of pain.

**Comment by ev-br** [comment]
Also note that the improved version  of the gist is now available at this link:

https://scipy.github.io/devdocs/notebooks/interp_transition_guide.html

Changing the reference from the gist to an official scipy tutorial chapter would be great actually.

**Comment by j-bowhay** [comment]
@ev-br maybe this is the answer to whether we press the button on #18618. I am happy to instead close it and just update the link?

**Comment by ev-br** [comment]
@j-bowhay agreed, updating the link and re-milestoning gh-18618 to 1.13 sounds great.

**Comment by h-vetinari** [comment]
> @j-bowhay agreed, updating the link and re-milestoning [gh-18618](https://github.com/scipy/scipy/pull/18618) to 1.13 sounds great.

There's now an issue to track the deprecation process around interp2d: #18583, perhaps good to move discussion there, or at least record the decisions you've come to?

**Comment by kcharlie2** [comment]
I would love to see this. n-dimensional interpolation with PChip and other splines has great potential, but the performance of those is 2-3 orders of magnitude greater than linear.

I was also able to get a modest speed improvement for all RegularGridInterpolator methods by simple code cleanup, for example caching more data and using array methods instead of numpy functions (e.g. array.all() instead of np.all(array)).

**Comment by ev-br** [comment]
> I was also able to get a modest speed improvement for all RegularGridInterpolator methods by simple code cleanup, for example caching more data and using array methods instead of numpy functions (e.g. array.all() instead of np.all(array)).

Would you be interested in submitting your changes as a PR for us to look at?
